tests/base.py

```diff
@@ -0,0 +1,422 @@
+                for message in output_file[stream_name]['messages']
+                if message['action'] == 'upsert']
+
+    def get_unique_records(self, stream, records):
+        unique_records = set()
+        for record in records:
+            # Build the primary key for this record, maintaining the same order for the fields
+            record_primary_key = [record[field]
+                                  for field in sorted(self.expected_primary_keys()[stream])]
+            # Cast to a tuple to make it hashable
+            unique_records.add(tuple(record_primary_key))
+        return unique_records
+
+    def verify_is_selected(self, is_selected):
+        """This is the assertion for the metadata for each field for every stream"""
+        self.assertTrue(is_selected)
+
+    def verify_stream_and_field_selection(self, conn_id):
+        """
+        For expected sync streams, verify that
+        - fields that should be selected are selected
+          - controlled by `self.verify_is_selected()`
+        - automatic fields are automatic
+        - non-automatic fields are "inclusion": "available"
+
+        For streams we don't expect to sync, verify that
+        - "selected" is false
+        """
+        catalogs = menagerie.get_catalogs(conn_id)
+
+        for catalog_entry in catalogs:
+            tap_stream_id = catalog_entry['tap_stream_id']
+
+            with self.subTest(tap_stream_id=tap_stream_id):
+                schema = menagerie.get_annotated_schema(conn_id, catalog_entry['stream_id'])
+                entry_metadata = schema.get('metadata', [])
+                if tap_stream_id in self.expected_sync_streams():
+                    for mdata in entry_metadata:
+                        is_selected = mdata.get('metadata').get('selected')
+                        if mdata.get('breadcrumb') == []:
+                            # Verify all expected sync streams are selected
+                            self.assertTrue(is_selected)
+                        else:
+                            inclusion = mdata.get('metadata', {}).get('inclusion')
+                            field_name = mdata.get('breadcrumb', ['properties', None])[1]
+
+                            automatic_fields = self.expected_automatic_fields()[tap_stream_id]
+
+                            self.assertIsNotNone(field_name)
+
+                            if field_name in automatic_fields:
+                                self.assertTrue(inclusion == 'automatic')
+                            else:
+                                self.verify_is_selected(is_selected)
+                                self.assertTrue(inclusion == 'available')
+                else:
+                    for mdata in entry_metadata:
+                        if mdata.get('breadcrumb') == []:
+                            self.assertFalse(mdata.get('metadata').get('selected'))
+
+    def make_connection_and_run_sync(self, create_connection_kwargs=None, selection_kwargs=None):
```

Comment:
I'd prefer if we used the existing common action methods. I think this is abstracting away too much. But I am open to discussing it.